core/peerset/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Peer Set Manager (PSM). Contains the strategy for choosing which nodes the network should be
+//! connected to.
+
+use std::collections::HashSet;
+use futures::{prelude::*, sync::mpsc};
+use libp2p::PeerId;
+use parking_lot::Mutex;
+use std::sync::Arc;
+
+/// Shared part of the peer set manager (PSM). Distributed around the code.
+pub struct Peerset {
+	tx: mpsc::UnboundedSender<Message>,
+	inner: Mutex<Inner>,
```

Comment:
Could you keep this `Inner` hidden in one particular future/stream(maybe `Behavior`?), and mutate it in response to receiving messages? 

If I understand it correctly each "operation" on `Peerset` first mutates `Inner` in some way, then sends a message to be handled by `Behavior`. Could we remove the `Mutex` by just keeping `Inner` inside `Behavior` and do the state mutation when receiving the message?

Same thing for `PeersetMut`, which mutates the same `Inner` and then sends a message that will be handled by `Behavior`. 

I'de rather have more state inside the messages, and the mutation of `Inner` being done at the point of handling the messages, which would appear to be inside `Behavior`. That way the mutex could be removed, and you'd get a more simple model of transition of the inner state of the `Peerset`. 

If other tasks or thread need to get updates on what's happening with `Inner`(which I haven't seen yet), this could be done through a stream of updates, as opposed to an `Arc`.

Reply:
> I'de rather have more state inside the messages, and the mutation of Inner being done at the point of handling the messages, which would appear to be inside Behavior. That way the mutex could be removed, and you'd get a more simple model of transition of the inner state of the Peerset.

That's a good point. This mutex is not necessary at all. But we can keep the `Inner` in the `PeersetMut`, there's no need to move it to the `Behaviour`.

The `PeersetMut` should be seen as "the thing that allocates slots", and it needs `Inner` for that.

On the other hand, this is a dummy implementation so I went for the straight-forward solution.

Reply:
Considering that the peerset is a dummy implementation, and considering that removing this mutex is purely an optimization, can I play the "won't fix" card on this?